packages/vite/src/node/optimizer/optimizer.ts

```diff
@@ -719,32 +719,48 @@ async function createDepsOptimizer(
     }
   }
 
-  function runOptimizerWhenIdle() {
-    if (!waitingOn) {
-      const next = registeredIds.pop()
-      if (next) {
-        waitingOn = next.id
-        const afterLoad = () => {
-          waitingOn = undefined
-          if (!closed && !workersSources.has(next.id)) {
-            if (registeredIds.length > 0) {
-              runOptimizerWhenIdle()
-            } else {
-              onCrawlEnd()
-            }
-          }
-        }
-        next
-          .done()
-          .then(() => {
-            setTimeout(
-              afterLoad,
-              registeredIds.length > 0 ? 0 : runOptimizerIfIdleAfterMs,
-            )
-          })
-          .catch(afterLoad)
+  async function runOptimizerWhenIdle() {
+    if (waitingOn.size > 0) return
+
+    const processingRegisteredIds = registeredIds
+    registeredIds = []
+
+    const donePromises = processingRegisteredIds.map(async (registeredId) => {
+      waitingOn.add(registeredId.id)
+      try {
+        await registeredId.done()
+      } finally {
+        waitingOn.delete(registeredId.id)
+      }
+    })
+
+    const afterLoad = () => {
+      if (closed) return
+      if (
+        registeredIds.length > 0 &&
+        registeredIds.every((registeredId) =>
+          workersSources.has(registeredId.id),
+        )
+      ) {
+        return
+      }
+
+      if (registeredIds.length > 0) {
+        runOptimizerWhenIdle()
+      } else {
+        onCrawlEnd()
       }
     }
+
+    const results = await Promise.allSettled(donePromises)
+    if (
+      registeredIds.length > 0 ||
+      results.some((result) => result.status === 'rejected')
+    ) {
+      afterLoad()
```

Comment:
The setTimeout 0 here was added at the time because things were failing if we did a direct call. It may have been a bug that prevented us to do it. I think we can try it out without it now, let's keep in mind that we could add it later here if needed (I don't think it would affect perf)